### PR TITLE
Refactor of the proto description logic for safety

### LINKED ISSUE #322: Unsafe mutations of shared state
**Describe the bug**
When creating a new cel.Env with custom protocol buffer types, the code mutates shared state without any locking.

**To Reproduce**
Discovered in thread sanitizer testing with multiple concurrent calls to

```
cel.NewEnv(cel.Types(&some.Message{}))
```

Other examples of mutation of shared state might exist and a thorough search or static analysis might be appropriate.  E.g. separately-created environments share underlying state via `pb.Db` from the `DefaultDb`.

Check which components this affects:

- [X]  parser
- [X]  checker
- [X]  interpreter

**Expected behavior**
Shared state should be immutable or changed only under a mutex.

**Additional context**
Seems to have been introduced in #291 in line 135 of `pb.go`.

### PR DESCRIPTION
**Pull request by TristonianJones**
Improvements in the initialization of `pb.FileDescription` and associated `pb.TypeDescription` values.

Notably, the `atomic.Value` members in the `pb.TypeDescription` have been replaced with a `sync.Once` object as the mutations were thread-safe, but should really only have been happening exactly once. 

Additionally, the reverse qualified name to file descriptor lookup is being managed entirely within the `pb.Db` as opposed to spread between the `pb.FileDescription` and `pb.Db` files. The net effect should be safer, simpler code that is better documented and better behaved.

Closes #322

### CODE REVIEW

File: common/types/pb/file.go [hunk 1]
```diff
@@ -70,44 +85,71 @@ func (fd *FileDescription) GetTypeNames() []string {
...
-	for _, enumType := range enumTypes {
-		for _, enumValue := range enumType.Value {
-			enumValueName := fmt.Sprintf(
-				"%s.%s.%s", pkg, enumType.GetName(), enumValue.GetName())
-			fd.enums[enumValueName] = &EnumDescription{
-				enumName: enumValueName,
-				file:     fd,
-				desc:     enumValue}
-			fd.pbdb.revFileDescriptorMap[enumValueName] = fd
+// fileMetadata is a flattened view of  message types and enum values within a file descriptor.
```
Comment (JimLarson) [inline common/types/pb/file.go]: Clarify whether the map keys are fully-qualified.

File: common/types/pb/enum.go [hunk 2]
```diff
@@ -18,19 +18,27 @@ import (
 	descpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
 )
 
-// EnumDescription maps a qualified enum name to its numeric value.
-type EnumDescription struct {
-	enumName string
-	file     *FileDescription
-	desc     *descpb.EnumValueDescriptorProto
+// NewEnumValueDescription produces a new named enum value description.
```
Comment (JimLarson) [inline common/types/pb/enum.go]: Clarify whether the name is fully qualified.

File: common/types/pb/enum.go [hunk 3]
```diff
@@ -18,19 +18,27 @@ import (
...
-	return ed.enumName
+// EnumValueDescription maps a qualified enum name to its numeric value.
+type EnumValueDescription struct {
+	enumValueName string
+	desc          *descpb.EnumValueDescriptorProto
 }
 
-// Value (numeric) of the enum.
-func (ed *EnumDescription) Value() int32 {
+// Name returns the human-readable identifier name for the enum value.
```
Comment (JimLarson) [inline common/types/pb/enum.go]: Fully qualified?

File: common/types/pb/file.go [hunk 4]
```diff
@@ -70,44 +85,71 @@ func (fd *FileDescription) GetTypeNames() []string {
...
-			file:     fd,
-			desc:     msgType,
-			metadata: &atomic.Value{},
+		msgName := fmt.Sprintf("%s.%s", container, msgType.GetName())
+		msgTypeMap[msgName] = msgType
+		nestedTypes := msgType.GetNestedType()
+		if len(nestedTypes) == 0 {
+			continue
+		}
+		nestedTypeMap := collectMsgTypes(msgName, nestedTypes)
```
Comment (JimLarson) [inline common/types/pb/file.go]: Consider making the type map an argument to `collectMsgTypes()` so it doesn't have to return a throwaway map.  Lines 132-139 could just be:

```
collectMsgTypes(msgName, msgType.GetNestedType(), msgTypeMap)
```

File: common/types/pb/file.go [hunk 5]
```diff
@@ -70,44 +85,71 @@ func (fd *FileDescription) GetTypeNames() []string {
...
+func collectFileMetadata(fileDesc *descpb.FileDescriptorProto) *fileMetadata {
+	pkg := fileDesc.GetPackage()
+	msgTypes := collectMsgTypes(pkg, fileDesc.GetMessageType())
+	enumValues := collectEnumValues(pkg, fileDesc.GetEnumType())
+	for container, msgType := range msgTypes {
+		nestedEnums := msgType.GetEnumType()
+		if len(nestedEnums) == 0 {
+			continue
+		}
+		nestedEnumValues := collectEnumValues(container, nestedEnums)
```
Comment (JimLarson) [inline common/types/pb/file.go]: Consider making the enum value map an argument to `collectEnumValue()` so it doesn't have to return a throwaway map.  Then the body of this loop could just be

```
collectEnumValue(container, msgType.getEnumType(), enumValueMap)
```

File: common/types/pb/type.go [hunk 6]
```diff
@@ -18,23 +18,50 @@ import (
...
-	"sync/atomic"
+	"sync"
 
 	"github.com/golang/protobuf/proto"
 	descpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
 	structpb "github.com/golang/protobuf/ptypes/struct"
 	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 )
 
+// NewTypeDescription produces a TypeDescription value for the proto type name with a given
```
Comment (JimLarson) [inline common/types/pb/type.go]: Specify whether the type name is fully-qualified.

**Review by JimLarson — APPROVED** [review]
LGTM, but with requests for clarification in comments whether names are fully-qualified, and a suggestion for simplification.

File: common/types/pb/file.go (hunk 1 above)
Comment (TristonianJones) [reply common/types/pb/file.go]: Done.

File: common/types/pb/enum.go (hunk 2 above)
Comment (TristonianJones) [reply common/types/pb/enum.go]: Done.

File: common/types/pb/enum.go (hunk 3 above)
Comment (TristonianJones) [reply common/types/pb/enum.go]: Done.

File: common/types/pb/file.go (hunk 4 above)
Comment (TristonianJones) [reply common/types/pb/file.go]: Done.

File: common/types/pb/file.go (hunk 5 above)
Comment (TristonianJones) [reply common/types/pb/file.go]: Done.

File: common/types/pb/type.go (hunk 6 above)
Comment (TristonianJones) [reply common/types/pb/type.go]: Done.

**Pull request merged**